frontends/parsers/parserDriver.cpp

```diff
@@ -0,0 +1,238 @@
+#include "parserDriver.h"
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/format.hpp>
+
+#include <cerrno>
+#include <cstdio>
+#include <iostream>
+#include <sstream>
+
+#include "frontends/common/constantFolding.h"
+#include "frontends/parsers/p4/p4lexer.hpp"
+#include "frontends/parsers/p4/p4parser.hpp"
+#include "frontends/parsers/v1/v1lexer.hpp"
+#include "frontends/parsers/v1/v1parser.hpp"
+#include "lib/error.h"
+
+
+#ifdef HAVE_LIBBOOST_IOSTREAMS
+
+#include <boost/iostreams/device/file_descriptor.hpp>
+#include <boost/iostreams/stream.hpp>
+
+namespace {
+
+/// A RAII helper class that provides an istream wrapper for a stdio FILE*. This
+/// is the efficient implementation for users with boost::iostreams installed.
+struct AutoStdioInputStream {
+    explicit AutoStdioInputStream(FILE* in)
+        : source(fileno(in), boost::iostreams::never_close_handle)
+        , buffer(source)
+        , stream(&buffer)
+    { }
+
+    std::istream& get() { return stream; }
+
+ private:
+    AutoStdioInputStream(const AutoStdioInputStream&) = delete;
+    AutoStdioInputStream(AutoStdioInputStream&&) = delete;
+
+    boost::iostreams::file_descriptor_source source;
+    boost::iostreams::stream_buffer<boost::iostreams::file_descriptor_source> buffer;
+    std::istream stream;
+};
+
+}  // anonymous namespace
+
+#else
+
+/// A RAII helper class that provides an istream wrapper for a stdio FILE*. This
+/// is an inefficient fallback implementation.
```

Comment:
This implementation copies the entire file multiple times, but yeah, it's still not really a limiting factor. I'd rather use only the boost::iostreams implementation, though. Waste not, want not.